k2/csrc/ragged_test.cu

```diff
@@ -1872,4 +1872,72 @@ TEST(RaggedShapeOpsTest, TestMakeTransposable) {
   TestMakeTransposable<kCuda>();
 }
 
+TEST(RaggedShapeDim, Clone) {
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    Array1<int32_t> row_splits(context, std::vector<int32_t>{0, 1, 2});
+    Array1<int32_t> row_ids(context, std::vector<int32_t>{0, 1});
+
+    RaggedShapeDim shape_dim;
+    shape_dim.row_splits = row_splits;
+    shape_dim.row_ids = row_ids;
+    shape_dim.cached_tot_size = 2;
+
+    RaggedShapeDim copied = shape_dim.Clone();
+    shape_dim.cached_tot_size += 1;
+    Increment(&shape_dim.row_splits);
+    Increment(&shape_dim.row_ids);
+
+    for (int32_t i = 0; i != shape_dim.row_splits.Dim(); ++i)
+      EXPECT_EQ(copied.row_splits[i] + 1, shape_dim.row_splits[i]);
+
+    for (int32_t i = 0; i != shape_dim.row_ids.Dim(); ++i)
+      EXPECT_EQ(copied.row_ids[i] + 1, shape_dim.row_ids[i]);
+
+    EXPECT_EQ(copied.cached_tot_size + 1, shape_dim.cached_tot_size);
+  }
+}
+
+TEST(RaggedShape, Clone) {
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    Array1<int32_t> row_splits(context, std::vector<int32_t>{0, 1, 2});
+    Array1<int32_t> row_ids(context, std::vector<int32_t>{0, 1});
+    RaggedShape shape = RaggedShape2(&row_splits, &row_ids, 2);
+    RaggedShape copied = shape.Clone();
+    EXPECT_TRUE(context->IsCompatible(*shape.Context()));
+    EXPECT_EQ(copied.NumAxes(), 2);
+    Increment(&shape.RowSplits(1));
+    Increment(&shape.RowIds(1));
+
+    for (int32_t i = 0; i != shape.RowSplits(1).Dim(); ++i)
+      EXPECT_EQ(copied.RowSplits(1)[i] + 1, shape.RowSplits(1)[i]);
+
+    for (int32_t i = 0; i != shape.RowIds(1).Dim(); ++i)
+      EXPECT_EQ(copied.RowIds(1)[i] + 1, shape.RowIds(1)[i]);
+  }
+}
+
+TEST(Ragged, Clone) {
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    Array1<int32_t> row_splits(context, std::vector<int32_t>{0, 1, 2});
+    Array1<int32_t> row_ids(context, std::vector<int32_t>{0, 1});
+    RaggedShape shape = RaggedShape2(&row_splits, &row_ids, 2);
+    Array1<int32_t> values(context, std::vector<int32_t>{10, 20});
+    Ragged<int32_t> ragged(shape, values);
+
+    Ragged<int32_t> copied = ragged.Clone();
+    Increment(&ragged.RowSplits(1));
+    Increment(&ragged.RowIds(1));
+    Increment(&ragged.values);
+
+    for (int32_t i = 0; i != ragged.RowSplits(1).Dim(); ++i)
+      EXPECT_EQ(copied.RowSplits(1)[i] + 1, ragged.RowSplits(1)[i]);
```

Comment:
it's OK to user `operator[]` on a small array, but it's better if we can notice it as it may call multiple device-to-host-copy in a loop. It's relatively OK here as the array is not so long, but we definitely need to convert to cpu first for a long array.

Reply:
Yes, you're right. This is only for unittests.

Reply:
I mean, even for unintests we should try to avoid copy too many times (RE a long array). Anyway, it's OK here as this array just has a small size.